### PR TITLE
Aggregation-in-order from query plan

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `aggregation-in-order` optimization on top of query plan. It is enabled by default (but works only together with `optimize_aggregation_in_order`, which is disabled by default). Set `query_plan_aggregation_in_order = 0` to use previous AST-based version.

### CODE REVIEW

File: src/Processors/QueryPlan/Optimizations/enableMemoryBoundMerging.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            reading->enforceAggregationInOrder();
+    }
+
+    for (auto & reading : async_reading_steps)
+    {
+        reading->enforceSorting(sort_description);
+        if (enforce_aggregation_in_order)
+            reading->enforceAggregationInOrder();
+    }
+
```
Comment (nickitat) [inline src/Processors/QueryPlan/Optimizations/enableMemoryBoundMerging.cpp:90]: would be good also to call `updateInputStream` on parent steps to propagate sorting

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp [hunk 2]
```diff
@@ -679,6 +679,212 @@ InputOrderInfoPtr buildInputOrderInfo(
...
+/// We really need three different sort descriptions here.
+/// For example:
+///
+///   create table tab (a Int32, b Int32, c Int32, d Int32) engine = MergeTree order by (a, b, c);
+///   select a, any(b), c, d from tab where b = 1 group by a, c, d order by c, d;
+///
+/// We would like to have:
+/// (a, b, c) - a sort description for reading from table (it's into input_order)
+/// (a, c) - a sort description for merging (an input of AggregatingInOrderTransfrom is sorted by this GROUP BY keys)
+/// (a, c, d) - a group by soer description (an input of FinishAggregatingInOrderTransform is sorted by all GROUP BY keys)
```
Comment (nickitat) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:691]:
```suggestion
/// (a, c, d) - a group by sort description (an input of FinishAggregatingInOrderTransform is sorted by all GROUP BY keys)
```

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp [hunk 3]
```diff
@@ -679,6 +679,212 @@ InputOrderInfoPtr buildInputOrderInfo(
...
+    const FixedColumns & fixed_columns,
+    const ActionsDAGPtr & dag,
+    const Names & group_by_keys,
+    const ActionsDAG & sorting_key_dag,
+    const Names & sorting_key_columns)
+{
+    MatchedTrees::Matches matches;
+    FixedColumns fixed_key_columns;
+
+    /// For every column in PK find any macth from GROUP BY key.
```
Comment (nickitat) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:712]:
```suggestion
    /// For every column in PK find any match from GROUP BY key.
```

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp [hunk 4]
```diff
@@ -679,6 +679,212 @@ InputOrderInfoPtr buildInputOrderInfo(
...
+    ReverseMatches reverse_matches;
+
+    if (dag)
+    {
+        matches = matchTrees(sorting_key_dag, *dag);
+
+        for (const auto & [node, match] : matches)
+        {
+            //std::cerr << "------- matching " << static_cast<const void *>(node) << " " << node->result_name
+            //    << " to " << static_cast<const void *>(match.node) << " " << (match.node ? match.node->result_name : "") << std::endl;
```
Comment (nickitat) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:723]:
```suggestion
```

File: src/Processors/QueryPlan/ReadFromRemote.cpp [hunk 5]
```diff
@@ -51,6 +52,32 @@ static void addConvertingActions(Pipe & pipe, const Block & header)
...
+}
+
+static void enforceAggregationInOrder(QueryProcessingStage::Enum stage, Context & context)
+{
+    if (stage != QueryProcessingStage::WithMergeableState)
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Cannot enforce aggregation in order for ReadFromRemote step up to stage {}",
+            QueryProcessingStage::toString(stage));
+
```
Comment (nickitat) [inline src/Processors/QueryPlan/ReadFromRemote.cpp:76]:
```suggestion

context.setSetting("enable_memory_bound_merging_of_aggregation_results", true);
```

**Comment by nickitat** [comment]
maybe relevant slowdowns:
https://s3.amazonaws.com/clickhouse-test-reports/43592/5edc9953b0b303a8be0a82b9df54c09abf77bb07/performance_comparison_[1/4]/report.html
https://s3.amazonaws.com/clickhouse-test-reports/43592/5edc9953b0b303a8be0a82b9df54c09abf77bb07/performance_comparison_[3/4]/report.html

File: src/Processors/QueryPlan/Optimizations/enableMemoryBoundMerging.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/enableMemoryBoundMerging.cpp:90]: Good idea.
Don't want to do it is this pr.

File: src/Processors/QueryPlan/ReadFromRemote.cpp (hunk 5 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/ReadFromRemote.cpp:76]: I would rather separate enforceAggregationInOrder and enforceSorting calls.

**Comment by KochetovNicolai** [comment]
Checked query from perftest and don't see any difference:
```
--- New
localhost:9000, queries 114, QPS: 0.363, RPS: 36284788.405, MiB/s: 276.831, result RPS: 0.000, result MiB/s: 0.000.

0.000%		2.556 sec.	
10.000%		2.629 sec.	
20.000%		2.673 sec.	
30.000%		2.692 sec.	
40.000%		2.721 sec.	
50.000%		2.745 sec.	
60.000%		2.765 sec.	
70.000%		2.783 sec.	
80.000%		2.811 sec.	
90.000%		2.889 sec.	
95.000%		2.956 sec.	
99.000%		3.108 sec.	
99.900%		3.125 sec.	
99.990%		3.125 sec.	

--- Old
localhost:9000, queries 242, QPS: 0.367, RPS: 36722549.794, MiB/s: 280.171, result RPS: 0.000, result MiB/s: 0.000.

0.000%		2.530 sec.	
10.000%		2.614 sec.	
20.000%		2.648 sec.	
30.000%		2.668 sec.	
40.000%		2.688 sec.	
50.000%		2.710 sec.	
60.000%		2.735 sec.	
70.000%		2.751 sec.	
80.000%		2.778 sec.	
90.000%		2.818 sec.	
95.000%		2.883 sec.	
99.000%		3.128 sec.	
99.900%		3.203 sec.	
99.990%		3.203 sec.
```

**Comment by KochetovNicolai** [comment]
Merged cause only comments were changed since last checks.

**Pull request merged**